### PR TITLE
Add support for menu subsubitems

### PR DESCRIPTION
**Pull request by jomunker**
depends on #1233

This PR adds support for menu subsubitems (third level items).

## Screenshots

image
image

### CODE REVIEW

**Comment by jomunker** [comment]
- [ ] add changeset

File: packages/admin/admin/src/mui/menu/CollapsibleItem.tsx [hunk 1]
```diff
@@ -53,7 +53,7 @@ const CollapsibleItem: React.FC<WithStyles<typeof styles> & MenuCollapsibleItemP
     children,
     ...otherProps
 }) => {
-    const itemLevel: 1 | 2 = level ? level : 1;
+    const itemLevel: MenuLevel["level"] = level ? level : 1;
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/CollapsibleItem.tsx:56]: 👍🏼

File: packages/admin/admin/src/mui/menu/Item.styles.ts [hunk 2]
```diff
@@ -0,0 +1,177 @@
+import { Theme } from "@mui/material";
+import { createStyles } from "@mui/styles";
+
+import { MenuItemProps, MuiListItemProps } from "./Item";
+
+export type MenuItemClassKey = "root" | "level1" | "level2" | "level3" | "hasIcon" | "hasSecondaryText" | "hasSecondaryAction";
+
+const colors = {
+    textLevel1: "#242424",
+    textLevel2: "#17181A",
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/Item.styles.ts:10]: These should come from the theme

File: packages/admin/admin/src/mui/menu/Item.styles.ts [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        },
+        level2: {
+            color: colors.textLevel2,
+            paddingLeft: 20,
+            paddingRight: 20,
+            paddingTop: 10,
+            paddingBottom: 10,
+            "&:last-child": {
+                borderBottom: ({ level, hasChildElements, isCollapsibleOpen }) =>
+                    level === 2 && (!hasChildElements || !isCollapsibleOpen) ? `1px solid ${theme.palette.grey[50]}` : "initial",
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/Item.styles.ts:68]: For what do you need this?

File: packages/admin/admin/src/mui/menu/CollapsibleItem.tsx [hunk 4]
```diff
@@ -82,6 +82,8 @@ const CollapsibleItem: React.FC<WithStyles<typeof styles> & MenuCollapsibleItemP
                     primary={primary}
                     showText={showText}
                     secondary={secondary}
+                    hasChildElements={!!childElements?.length}
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/CollapsibleItem.tsx:85]: How about `hasSubitems`?

File: packages/admin/admin/src/mui/menu/Item.styles.ts (hunk 3 above)
Comment (jomunker) [reply packages/admin/admin/src/mui/menu/Item.styles.ts:68]: To have the border move from this:
Screenshot 2023-12-22 at 09 37 57
to this:
Screenshot 2023-12-22 at 09 38 06

File: packages/admin/admin/src/mui/menu/Item.tsx [hunk 5]
```diff
@@ -118,14 +28,15 @@ const Item: React.FC<WithStyles<typeof styles> & MenuItemProps & MuiListItemProp
 }) => {
     const context = React.useContext(MenuContext);
     if (!context) throw new Error("Could not find context for menu");
-    if (level > 2) throw new Error("Maximum nesting level of 2 exceeded.");
+    if (level > 3) throw new Error("Maximum nesting level of 2 exceeded.");
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/Item.tsx:31]:
```suggestion
    if (level > 3) throw new Error("Maximum nesting level of 3 exceeded.");
```

File: packages/admin/admin/src/mui/menu/Item.tsx [hunk 6]
```diff
@@ -1,111 +1,21 @@
...
-    });
+import { MenuItemClassKey, styles } from "./Item.styles";
 
 export interface MenuItemProps extends MenuLevel {
     primary: React.ReactNode;
     secondary?: React.ReactNode;
     icon?: React.ReactElement;
     secondaryAction?: React.ReactNode;
+    isCollapsibleOpen?: boolean;
+    hasSubitems?: boolean;
```
Comment (johnnyomair) [inline packages/admin/admin/src/mui/menu/Item.tsx:15]: Why do you need this two states? Which styling does this affect?

File: packages/admin/admin/src/mui/menu/Item.tsx (hunk 5 above)
Comment (jomunker) [reply packages/admin/admin/src/mui/menu/Item.tsx:31]: 2 times is correct. level 1 is root and you can nest two times ;)

File: packages/admin/admin/src/mui/menu/Item.tsx (hunk 6 above)
Comment (jomunker) [reply packages/admin/admin/src/mui/menu/Item.tsx:15]: I need it in the .styles file for the border

Comment (jomunker) [reply packages/admin/admin/src/mui/menu/Item.tsx:15]: image

**Comment by johnnyomair** [comment]
I'll review this once #1233 has been merged. Setting as draft for now.

**Comment by johnnyomair** [comment]
@jomunker the base PR has been merged, please update this branch and solve conflicts so we can review it again, thanks!

**Review by jamesricky — CHANGES_REQUESTED** [review]
Could you please add an example of this feature to the dev-stories?
Either in a new story or maybe in the existing menu-story: `packages/admin/admin-stories/src/admin/mui/Menu.tsx` 

This would be very helpful for review and for future changes/debugging of this feature.

**Comment by jamesricky** [comment]
@jomunker the implementation seems to differ from the design:
- The selected sub-sub items should not have a blue background but the text and the border on the left should be blue instead. 
- The left border of the last item should extend all the way to the bottom of the item, like the border on the first item extends all the way to the top. 

Are these things already implemented in one of the follow-up PRs? 
Or do I maybe have an outdated design?

| Current implementation | Design |
| ---------- | --- |
| Current implementation | Design |

File: demo/site/src/theme.ts [hunk 7]
```diff
@@ -78,7 +78,7 @@ const theme: Theme = {
         n500: "#4C4C4C",
         n600: "#404040",
         n700: "#333333",
-        n800: "#242424",
+        n800: "#24242gl" + "4",
```
[the remark this answers is not in the document]
Comment (johnnyomair) [reply demo/site/src/theme.ts:81]: I don't think so 😂

File: packages/api/cms-api/schema.gql [hunk 8]
```diff
@@ -123,6 +123,11 @@ type FilenameResponse {
   isOccupied: Boolean!
 }
 
+type SitePreviewHash {
+  timestamp: Float!
+  hash: String!
+}
+
```
Comment (johnnyomair) [inline packages/api/cms-api/schema.gql:130]: This change seems unrelated to me 🤔

File: packages/admin/admin/src/index.ts [hunk 9]
```diff
@@ -103,7 +103,8 @@ export { MasterLayoutClassKey } from "./mui/MasterLayout.styles";
 export { MasterLayoutContext } from "./mui/MasterLayoutContext";
 export { MenuCollapsibleItem, MenuCollapsibleItemClassKey, MenuCollapsibleItemProps, MenuLevel } from "./mui/menu/CollapsibleItem";
 export { IMenuContext, IWithMenu, MenuContext, withMenu } from "./mui/menu/Context";
-export { MenuItem, MenuItemClassKey, MenuItemProps } from "./mui/menu/Item";
+export { MenuItem, MenuItemProps } from "./mui/menu/Item";
+export { MenuItemClassKey } from "./mui/menu/Item.styles";
```
Comment (johnnyomair) [inline packages/admin/admin/src/index.ts:107]: Note: the theming of components is currently refactored in #1376, removing .styles files and the `createStyles` usage. The theme refactor will presumably be finished before the menu rework, so you'll have to change the styles in this PR.

Comment (jamesricky) [reply packages/admin/admin/src/index.ts:107]: I think this should be done in a separate PR in the feature branch, once #1376 has been merged.

File: demo/site/src/theme.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jomunker) [reply demo/site/src/theme.ts:81]: haha i think this was a missplaced git pull (gl) + enter and an auto liniting afterwards :joy

File: packages/admin/admin/src/index.ts (hunk 9 above)
Comment (jomunker) [reply packages/admin/admin/src/index.ts:107]: Agree 👍🏻

File: packages/api/cms-api/schema.gql (hunk 8 above)
Comment (jomunker) [reply packages/api/cms-api/schema.gql:130]: hmm weird. Seems like a generator change?

Comment (jomunker) [reply packages/api/cms-api/schema.gql:130]: Should I revert it or just leave it because it basically just reordered it?

Comment (johnnyomair) [reply packages/api/cms-api/schema.gql:130]: This will be resolved after merging next into the feature branch, so we may ignore it here.

In general, it's good idea to not add changes which aren't related to your changes (i.e., no `git add .`)

**Comment by jomunker** [comment]
> @jomunker the implementation seems to differ from the design:
> 
> * The selected sub-sub items should not have a blue background but the text and the border on the left should be blue instead.
> * The left border of the last item should extend all the way to the bottom of the item, like the border on the first item extends all the way to the top.
> 
> Are these things already implemented in one of the follow-up PRs? Or do I maybe have an outdated design?
> 
> Current implementation	Design
> Current implementation	Design

Seems like i fixed the design in the follow up PR. See the following screen recording. Sorry for the confusion with the follow up PRs.

https://github.com/vivid-planet/comet/assets/56400587/f9d2343a-dbcc-4d44-9ca1-baa2ed97f856

File: packages/admin/admin/src/mui/menu/Item.styles.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        },
+        level2: {
+            color: colors.textLevel2,
+            paddingLeft: 20,
+            paddingRight: 20,
+            paddingTop: 10,
+            paddingBottom: 10,
+            "&:last-child": {
+                borderBottom: ({ level, hasSubitems, isCollapsibleOpen }) =>
+                    level === 2 && (!hasSubitems || !isCollapsibleOpen) ? `1px solid ${theme.palette.grey[50]}` : "initial",
```
[the remark this answers is not in the document]
Comment (jomunker) [reply packages/admin/admin/src/mui/menu/Item.styles.ts]: Good catch 👍🏻

**Comment by jamesricky** [comment]
> Seems like i fixed the design in the follow up PR. See the following screen recording.

The bottom part of the border on the last item is still missing in the video. 
Can you please add that? In a follow-up PR is fine. 

Screenshot 2024-02-29 at 15 58 24

**Review by johnnyomair — APPROVED** [review]
